### PR TITLE
enhance nodediscovery process: don't write "NOIP" if the node can not be resloved to an IP

### PR DESCRIPTION
**Pull request by zet809**
The fix for #1270

enhance_nodediscover_process, don't write mac table and don't generate dhcp lease entry

In the client, it will be in the sending findme request loop, after admin run "makehosts <pre-defined-node>", the hardware discovery can update MAC address successfully after send out the next findme request.

The /var/log/xcat/cluster.log will show this right now:
```
Nov 15 11:16:34 briggs01 xcat[47212]: xcat.discovery.aaadiscovery: (70:e2:84:14:28:0e) Got a discovery request, attempting to discover the node...
Nov 15 11:16:34 briggs01 xcat[47212]: DEBUG xcatd: call plugin <blade> to handle command <findme>
Nov 15 11:16:34 briggs01 xcat[47212]: xcat.discovery.blade: (70:e2:84:14:28:0e) Warning: Could not find any nodes using blade-based discovery
Nov 15 11:16:34 briggs01 xcat[47212]: DEBUG xcatd: call plugin <profilednodes> to handle command <findme>
Nov 15 11:16:34 briggs01 xcat[47212]: DEBUG xcatd: call plugin <seqdiscovery> to handle command <findme>
Nov 15 11:16:34 briggs01 xcat[47212]: DEBUG xcatd: call plugin <switch> to handle command <findme>
Nov 15 11:16:34 briggs01 xcat[47212]: xcat.discovery.switch: (70:e2:84:14:28:0e) Found node: mid05tor12cn05
Nov 15 11:16:34 briggs01 xcat[47212]: DEBUG xcatd: call plugin <nodediscover> to handle command <discovered>
Nov 15 11:16:35 briggs01 xcat[47212]: xcat.discovery.nodediscover: Can not resolve IP for the matching node:mid05tor12cn05. Make sure "makehosts" and "makedns" have been run for mid05tor12cn05.
Nov 15 11:16:35 briggs01 xcat[47212]: xcat.discovery.zzzdiscovery: (70:e2:84:14:28:0e) Failed for node discovery: Can not resolve IP for the matching node:mid05tor12cn05.
Nov 15 11:16:35 briggs01 xcat[47212]: xcat.discovery.zzzdiscovery: Notify 172.12.139.5 that its findme request has been processed
```

### CODE REVIEW

**Comment by whowutwut** [comment]
@zet809 Are you planning on refining the last message since the discovery has failed? OR that will come in another PR?  

Should we be adding a message in that xcat.discovery logs based on the  changes here? I'm not following the changes very well, the title says don't write NOIP if it can't resolve IP address. so what would the MAC address show if it can't resolve IP?

**Comment by zet809** [comment]
hi, @whowutwut , yes, you are right that the discovery is failed in the end. I need to modify it.

For the original situation, if the node can not be resolved to an IP address, there will be a "NOIP" appended to the mac address and then set "<mac_address!*NOIP*>" to the mac attribute of the pre-defined node. Then, when the CN restart network server, it won't be able to get a dynamic IP any more since the "*NOIP*" will cause a "allow-booting = 00" entry in the dhcp lease file.
So, in this fix, we will report the error and don't set "NOIP" entry in order not to cause the CN failed to get a dynamic again.

**Comment by whowutwut** [comment]
@zet809  Ok, so the NOIP was actually helping us debug before, now with this change you're saying it caused issues unless we remove the NOIP from the mac string? So to avoid that, we just do not add it?

File: xCAT-server/lib/xcat/plugins/nodediscover.pm [hunk 1]
```diff
@@ -349,6 +351,9 @@ sub process_request {
                 if ($forcenic == 1) { $macstring .= $currmac . "|"; }
             }
         }
+        if ($macs_withip == 1 and $macstring =~ /\*NOIP\*/) {
+            return;
```
Comment (robin2008) [inline xCAT-server/lib/xcat/plugins/nodediscover.pm]: Why not just remove `*NOIP*` here, but continue with the other parts in this function.
I see lots of logic there, for example, makedhcp, updateswitch, etc.

**Comment by zet809** [comment]
Hi, @whowutwut and @robin2008 , I have update the PR, will you pls help to take a look at again? Thx!

**Comment by immarvin** [comment]
i am ok with the change

**Pull request merged**